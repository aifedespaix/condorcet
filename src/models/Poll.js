export const PollModel = {
  answers: "required|array|min:2",
  "answers.*.slug": "required|string",
  "answers.*.value": "required|string",
  description: "string",
  isActive: "required|boolean",
  isPublic: "required|boolean",
  owner: "required|string",
  question: "required|string",
  token: "required|string",
  users: "required|array",
  timestamp: "object"
};
