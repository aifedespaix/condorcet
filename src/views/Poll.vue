<template>
  <div>
    <p class="text-center">
      Pour voter il suffit de classer vos préférences en les glissant-déposant :
    </p>

    <SortableList
      class="mt-6 mx-auto w-full px-2"
      lockAxis="y"
      v-model="answers"
      @input="onMove"
    >
      <SortableItem
        v-for="(answer, index) in answers"
        :index="index"
        :key="answer.rank"
        :rank="answer.rank"
        :value="answer.value"
      ></SortableItem>
    </SortableList>

    <button
      class="mt-4 w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
      @click="vote"
    >
      Voter
    </button>
  </div>
</template>

<script>
import { mapState, mapActions, mapGetters } from "vuex";
import shuffle from "shuffle-array";
import SortableList from "../components/SortableList";
import SortableItem from "../components/SortableItem";
import { createVote } from "../utils/request";

export default {
  name: "Poll",
  components: {
    SortableList,
    SortableItem
  },
  data() {
    return {
      answers: []
    };
  },
  computed: {
    ...mapGetters("user", ["userId"]),
    ...mapState({
      poll: state => state.poll.current,
      userId: state => state.user.id
    })
  },
  async mounted() {
    await this.setPoll(this.$route.params.id);

    if (!this.poll.users.includes(this.userId)) {
      this.$router.push({ name: "polls_list" });
    }

    this.answers = shuffle(this.poll.answers).map((answer, index) => {
      return {
        rank: index + 1,
        ...answer
      };
    });
  },
  methods: {
    ...mapActions("poll", ["setPoll", "fetchPoll"]),
    ...mapActions("app", ["addNotification"]),
    onMove(inputs) {
      this.answers = inputs.map((input, index) => {
        return {
          ...input,
          rank: index + 1
        };
      });
    },
    async vote() {
      const pollId = this.$route.params.id;
      const userId = this.userId;

      try {
        await createVote({ vote: this.answers }, pollId, userId);

        this.addNotification({
          text: "Merci de ton vote",
          status: "success"
        });

        this.$router.push({
          name: "polls_list"
        });
      } catch (error) {
        this.addNotification({
          text: "Il y a eu un problème",
          status: "error"
        });
      }
    }
  }
};
</script>

<style></style>
