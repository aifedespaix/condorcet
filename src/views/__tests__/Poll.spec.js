import { shallowMount } from "@vue/test-utils";
import Poll from "../Poll.vue";

const $route = {
  params: "adz93dze823j"
};

describe("Poll.vue", () => {
  it("should render", () => {
    const wrapper = shallowMount(Poll, {
      mocks: {
        $route
      }
    });

    expect(wrapper).toMatchSnapshot();
  });
});
