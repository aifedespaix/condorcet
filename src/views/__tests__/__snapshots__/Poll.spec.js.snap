// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poll.vue should render 1`] = `
<div>
  <div>
    <p>
      Pour voter il suffit de classer vos préférences en les glissant-déposant
    </p>
    <v-layout-stub tag="div" row="">
      <v-flex-stub tag="div" xs12="" sm10="" offset-sm1="">
        <v-card-stub tag="div">
          <sortablelist-stub value="" axis="y" distance="0" pressdelay="0" pressthreshold="5" hidesortableghost="true" lockoffset="50%" transitionduration="300" appendto="body" lockaxis="y" shouldcancelstart="function _default(e) {
        // Cancel sorting if the event target is an \`input\`, \`textarea\`, \`select\` or \`option\`
        var disabledElements = ['input', 'textarea', 'select', 'option', 'button'];
        return disabledElements.indexOf(e.target.tagName.toLowerCase()) !== -1;
      }" gethelperdimensions="function _default(_ref) {
        var node = _ref.node;
        return {
          width: node.offsetWidth,
          height: node.offsetHeight
        };
      }"></sortablelist-stub>
        </v-card-stub>
      </v-flex-stub>
    </v-layout-stub>
    <v-btn-stub color="info" activeclass="v-btn--active" tag="button" type="button">Valider</v-btn-stub>
  </div>
</div>
`;
